io/src/main/scala/fs2/io/udp/AsynchronousSocketGroup.scala

```diff
@@ -0,0 +1,234 @@
+package fs2
+package io
+package udp
+
+import scala.collection.mutable.{Queue=>MutableQueue}
+
+import java.io.IOException
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+import java.nio.channels.{CancelledKeyException,ClosedChannelException,DatagramChannel,Selector,SelectionKey}
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * Supports read/write operations on an arbitrary number of UDP sockets using a shared selector thread.
+ *
+ * Each `AsynchronousSocketGroup` is assigned a single daemon thread that performs all read/write operations.
+ */
+sealed trait AsynchronousSocketGroup {
+  private[udp] type Context
+  private[udp] def register(channel: DatagramChannel): Context
+  private[udp] def read(ctx: Context, cb: Either[Throwable,Packet] => Unit): Unit
+  private[udp] def write(ctx: Context, packet: Packet, cb: Option[Throwable] => Unit): Unit
+  private[udp] def close(ctx: Context): Unit
+
+  /**
+   * Shuts down the daemon thread used for selection and rejects all future read/write requests.
+   */
+  def close: Unit
+}
+
+object AsynchronousSocketGroup {
+
+  def apply(): AsynchronousSocketGroup = new AsynchronousSocketGroup {
+
+    class Attachment(
+      readers: MutableQueue[Either[Throwable,Packet] => Unit] = MutableQueue(),
+      writers: MutableQueue[(Packet,Option[Throwable] => Unit)] = MutableQueue()
+    ) {
+
+      def dequeueReader: Option[Either[Throwable,Packet] => Unit] = {
+        if (readers.isEmpty) None
+        else Some(readers.dequeue())
+      }
+
+      def queueReader(reader: Either[Throwable,Packet] => Unit): Unit = {
+        if (closed) reader(Left(new ClosedChannelException))
+        else {
+          readers += reader
+          ()
+        }
+      }
+
+      def hasReaders: Boolean = readers.nonEmpty
+
+      def peekWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+        if (writers.isEmpty) None
+        else Some(writers.head)
+      }
+
+      def dequeueWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+        if (writers.isEmpty) None
+        else Some(writers.dequeue())
+      }
+
+      def queueWriter(writer: (Packet,Option[Throwable] => Unit)): Unit = {
+        if (closed) writer._2(Some(new ClosedChannelException))
```

Comment:
I originally had synchronized blocks on the methods of `Attachment` but I switched to a model where the attachment state is only read and written from the selector thread -- which is a basically a weak/informal actor model. If we wanted to run multiple selector threads in parallel, we'd need to reintroduce the synchronization and then replace `peekWriter` with `pushWriter` or something -- e.g., replace the pattern of `peek/dequeue` with `dequeue/push on failure`.

What do you think? I can change to that model if you think we might want to support multiple selector threads at some point in the future.